lsp-pwsh.el

```diff
@@ -0,0 +1,307 @@
+   ("powershell.codeFolding.showLastLine" lsp-pwsh-code-folding-show-last-line t)
+   ("powershell.codeFolding.enable" lsp-pwsh-code-folding-enable t)
+   ("powershell.scriptAnalysis.settingsPath" lsp-pwsh-script-analysis-settings-path)
+   ("powershell.scriptAnalysis.enable" lsp-pwsh-script-analysis-enable t)
+   ("powershell.helpCompletion" lsp-pwsh-help-completion)))
+
+;; lsp-pwsh custom variables
+(defcustom lsp-pwsh-ext-path (expand-file-name ".extension/pwsh"
+                                               user-emacs-directory)
+  "The path to powershell vscode extension."
+  :group 'lsp-pwsh
+  :type 'string)
+
+(defcustom lsp-pwsh-exe (or (executable-find "pwsh") (executable-find "powershell"))
+  "PowerShell executable."
+  :type 'string
+  :group 'lsp-pwsh)
+
+(defcustom lsp-pwsh-dir (expand-file-name "PowerShellEditorServices" lsp-pwsh-ext-path)
+  "Path to PowerShellEditorServices without last slash."
+  :type 'string
+  :group 'lsp-pwsh)
+
+(defvar lsp-pwsh-log-path (expand-file-name "logs" lsp-pwsh-ext-path)
+  "Path to directory where server will write log files.
+Must not nil.")
+
+(defvar lsp-pwsh--sess-id (emacs-pid))
+
+(defun lsp-pwsh--command ()
+  "Return the command to start server."
+  `(,lsp-pwsh-exe "-NoProfile" "-NonInteractive" "-NoLogo"
+                  ,@(if (eq system-type 'windows-nt) '("-ExecutionPolicy" "Bypass"))
+                  "-OutputFormat" "Text"
+                  "-File"
+                  ,(f-join lsp-pwsh-dir "PowerShellEditorServices/Start-EditorServices.ps1")
+                  "-HostName" "'Emacs Host'"
+                  "-HostProfileId" "'Emacs.LSP'"
+                  "-HostVersion" "0.1"
+                  "-LogPath" ,(f-join lsp-pwsh-log-path "emacs-powershell.log")
+                  "-LogLevel" ,lsp-pwsh-developer-editor-services-log-level
+                  "-SessionDetailsPath"
+                  ,(format "%s/PSES-VSCode-%d" lsp-pwsh-log-path lsp-pwsh--sess-id)
+                  ;; "-AdditionalModules" "@('PowerShellEditorServices.VSCode')"
+                  "-Stdio"
+                  "-BundledModulesPath" ,lsp-pwsh-dir
+                  "-FeatureFlags" "@()"
+                  ))
+
+(defun lsp-pwsh--extra-init-params ()
+  "Return form describing parameters for language server.")
+
+(defvar lsp-pwsh--major-modes '(powershell-mode))
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-stdio-connection #'lsp-pwsh--command)
+  :major-modes lsp-pwsh--major-modes
+  :server-id 'pwsh-ls
+  :priority 1
+  :initialization-options #'lsp-pwsh--extra-init-params
+  :notification-handlers (lsp-ht ("powerShell/executionStatusChanged" 'ignore))
+  :initialized-fn (lambda (w)
+                    (with-lsp-workspace w
+                      (lsp--set-configuration
+                       (lsp-configuration-section "powershell"))))
+  ))
+
+(add-to-list 'lsp-language-id-configuration '(powershell-mode . "powershell"))
+
+;; Compatibility
+(if (fboundp 'company-lsp)
```

Comment:
Are you able to make lsp-mode work with company-tng using this piece of code? Still, we have to move it somewhere.

Reply:
cc @dgutov - we were discussing the company-tng and this might be helpful.

Reply:
At this point I think this is kind of like language specific fix to get it works with other popular package. 
For language servers that suffer from `company-tng` is not expanding the post completion, I think PowerShellEditorService got hit the hardest. Since it actually erases part of text at beginning of completion candidate and need the post completion to get back.
I've already check if the advice is applied only for powershell though.
Is that good to still keep this here?

Reply:
> Is that good to still keep this here?

ATM company-tng is considered unsupported(or company-tng does not support company-lsp). IMO even if we submit it as it is we should work with upstream no fixing it.

In your experience, is this sufficient to make company-tng work with company-lsp? 

As a side note `(with-eval-after-load` might be a bit better than `fboundp`.

Reply:
> In your experience, is this sufficient to make company-tng work with company-lsp?

Yes, this's sufficient to make `company-tng` work with `lsp-pwsh`.

I really cannot say anything for other language servers, but in my experience it works acceptable.

Reply:
Merging as it is but I am reserving the right to reverse it in case of a problem.

Reply:
Great, thank you.

Reply:
@yyoncho I see the hack, it's... interesting. But if you have a specific suggestion how company-tng can be improved, I'm all ears.